### DISCUSSION TITLE
Co-relating snapshots with MQTT box values

### DISCUSSION THREAD

**Discussion by klutzzykitty** [blakeblackshear/frigate#4423] [Ask A Question]
First of all, thanks for the project. It's a great example of a real world application on merging AI and IOT. 

I'm trying to obtain positions of the detected object from its snapshots. 
Hence, before using the bounding box values from mqtt events topic, I check if it coincides with the actual bounding box on the snapshot of the detected object. 

Came across a similar discussion here: #3220

**Problem:** 
Most times the bounding box values from mqtt events topic doesn't coincide with the actual bounding box on the snapshot, which I guess should happen, isn't it ? 
Its only sometimes that the values match. 
Below are some example images. The blue rectangle (for reference) is the bounding box values from mqtt events topic for the corresponding snapshot, compared with the actual bounding around the detected object. 

**No match cases:**
![person2](https://user-images.githubusercontent.com/118447577/202482111-00306fd2-10c9-4204-8cab-81492b2264e0.png)
![person4](https://user-images.githubusercontent.com/118447577/202482147-d89bb166-6c37-401b-9657-47bf35c9938d.png)
(For reference), in the above image, the person doesn't go inside the fence in other timestamps of the video feed.
![nvperson1](https://user-images.githubusercontent.com/118447577/202482259-b19b5764-3c0d-493b-bf7d-6a87e3a662e8.png)

**Match case:**
![nvperson2](https://user-images.githubusercontent.com/118447577/202484526-9777e3ca-bede-40c9-ad20-00ee362e62bf.png)

**Method:** 
- Grabbing snapshots from HTTP api whenever it gets generated: `http://x.x.x.x:5000/api/events/eventid/snapshot.jpg?bbox=1&crop=1&quality=100	`
Not the ones that get saved int he `clips` folder. 
- Grabbing the box values from mqtt events topic: From every `after` section of the corresponding event to that snapshot. 
- Checking if the box values generated for the corresponding snapshot match with the bounding box in the snapshot so that the object's position can be concluded to be precise. 

Would this be the right way to do it ?

**Comment by NickM-27** [comment]
> Most times the bounding box values from mqtt events topic doesn't coincide with the actual bounding box on the snapshot, which I guess should happen, isn't it ?

No, the MQTT box is the real time location of the object. The snapshot is not the latest position of the object as the snapshots are only updated sometimes based on the detection score and some other factors.

**Reply by NickM-27** [reply]
You'll notice there's also a `snapshot_time` included in the payload which is when the snapshot was taken https://github.com/blakeblackshear/frigate/blob/dev/frigate/object_processing.py#L181

Also the decision for when to save an updated snapshot https://github.com/blakeblackshear/frigate/blob/dev/frigate/object_processing.py#L772

**Reply by NickM-27** [reply]
You could try using the `/api/<camera_name>/latest.jpg` and it should align with the object position better.

**Reply by klutzzykitty** [reply]
Thanks for that additional info @NickM-27 
Didn't know the published box values were real-time ones. That explains my misunderstanding. 

`/api/<camera_name>/latest.jpg` did in-fact align with the object's positions a bit better, but wasn't reliable as doing that would require 
processing `latest.jpg` which gets updated quickly depending on what frame of the video stream Frigate is processing. 

This would  add a computational overhead. 

What I was meddling with is more aligned towards processing the snapshot image from selective camera streams and determining the detected object's co-ordinates for further needs. This would reduce computations as it only gets triggered when a snapshot is generated from a particular camera in a set of cameras connected to Frigate instead of checking multiple frames form `/api/camera_name/latest.jpg`.

**Solution:** 
-  Conventional computer vision applied to the snapshot of the detected object. 
-  Determining the detected object's co-ordinates in the snapshot from image processing techniques. 
-  Isolating and extracting the detected object from the snapshot. 
-  Using the extracted object for further requirements, automation, extra details on the object etc.  

**Result:** 
Tried with the above image processing method and it seemed better suited to what I was trying to do.
